lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchTaskStatusAlarmHandler.java

```diff
@@ -0,0 +1,114 @@
+  private final InjectionFuture<AzureBatchResourceManager> azureBatchResourceManager;
+  private final InjectionFuture<REEFEventHandlers> reefEventHandlers;
+  private final int taskStatusCheckPeriod;
+  private boolean isAlarmScheduled;
+  private Clock clock;
+
+  private static final Logger LOG = Logger.getLogger(AzureBatchResourceLaunchHandler.class.getName());
+
+  @Inject
+  private AzureBatchTaskStatusAlarmHandler(
+      final InjectionFuture<REEFEventHandlers> reefEventHandlers,
+      final AzureBatchHelper azureBatchHelper,
+      final InjectionFuture<AzureBatchResourceManager> azureBatchResourceManager,
+      final InjectionFuture<AzureBatchResourceManagerStartHandler> azureBatchResourceManagerStartHandler,
+      @Parameter(AzureBatchTaskStatusCheckPeriod.class) final int taskStatusCheckPeriod,
+      final Clock clock) {
+    this.reefEventHandlers = reefEventHandlers;
+    this.azureBatchHelper = azureBatchHelper;
+    this.azureBatchResourceManager = azureBatchResourceManager;
+    this.azureBatchResourceManagerStartHandler = azureBatchResourceManagerStartHandler;
+    this.clock = clock;
+    this.taskStatusCheckPeriod = taskStatusCheckPeriod;
+  }
+
+  @Override
+  public void onNext(final Alarm alarm) {
+    String jobId = this.azureBatchHelper.getAzureBatchJobId();
+    List<CloudTask> allTasks = this.azureBatchHelper.getTaskStatusForJob(jobId);
+
+    if (this.isAlarmScheduled) {
+      this.scheduleAlarm();
+    }
+
+    // Report status if the task has an associated active container.
+    LOG.log(Level.FINER, "Found {0} tasks from job id {1}", new Object[]{allTasks.size(), jobId});
+    LOG.log(Level.FINEST, "active container list: {0}",
+        this.azureBatchResourceManager.get().activeContainerList());
+    for (CloudTask task : allTasks) {
+      State reefTaskState = TaskStatusMapper.getReefTaskState(task);
+      LOG.log(Level.FINEST, "status for Task Id: {0} is [Azure Batch Status]:{1}, [REEF status]:{2}",
+          new Object[]{task.id(), task.state().toString(), reefTaskState});
+      if (this.azureBatchResourceManager.get().isContainerActive(task.id())) {
+        ResourceStatusEvent resourceStatusEvent = ResourceStatusEventImpl.newBuilder()
+            .setIdentifier(task.id())
+            .setState(reefTaskState)
+            .build();
+        LOG.log(Level.FINEST, "Reporting status for Task Id: {0} is [Azure Batch Status]:{1}, [REEF status]:{2}",
+            new Object[]{task.id(), task.state().toString(), reefTaskState});
+        this.reefEventHandlers.get().onResourceStatus(resourceStatusEvent);
+      } else {
+        LOG.log(Level.FINEST,
+            "No active container in resource manager for Task: {0}. Not sending update.", task.id());
+      }
+    }
+  }
+
+  public void enableAlarm() {
+    this.isAlarmScheduled = true;
+    this.scheduleAlarm();
+  }
+
+  public void disableAlarm() {
+    this.isAlarmScheduled = false;
+  }
```

Comment:
Not sure if this is intended - by setting this, we only stop scheduling further alarm. Existing alarm will still fire.

If we want to stop the existing scheduled alarm, I think it is this.clock.stop()

Reply:
Oh no. I tried clock.stop() .. and that had the effect of keeping the driver running for ever and I had to manually terminate it. Since the clock is shared with other handlers, stopping the clock is not the right thing to do. Wont fix.

Reply:
I see. thanks!